Cleanup NativeItemExists()

iSazonov opened the pull request:
## PR Summary

Related #6874. After the PR we can remove the files - isfile.cpp, isfile.h, isdirectory.cpp, isdirectory.h - from psl.

- Use File.GetAttributes() instead of p/invokes. The method throws access exceptions (unlike File.Exists and Directory.Exists) which we use for better error messages.
- Remove 'Native' from method names.
- As stated in #6909 we get a regression in rare case (for files like pagefile.sys). Fix is ready in [CoreFX](https://github.com/dotnet/corefx/issues/29782), we get it in 2.1.1. I suggest ignore the regression because this is a very-very rare situation (Get-ChildItem c:\pagefile.sys -Hidden). The .Net Core team was not even able to create an artificial test for such files and uses a real pagefile.sys file for the test. Also the enumeration is still working (dir c:\ -hidden).

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

iSazonov committed:
Remove unneeded p/invokes

iSazonov committed:
Replace NativeItemExists in GetPathItems

iSazonov committed:
Replace NativeItemExists in RemoveDirectoryInfoItem

iSazonov committed:
Replace NativeItemExists in ItemExists

iSazonov committed:
Replace NativeItemExists in GetProperty

iSazonov committed:
Replace NativeItemExists in SetProperty

iSazonov committed:
Create NativeItemExists based on File.GetAttributes(path)

iSazonov committed:
Update NativeItemExists w/o throw

iSazonov committed:
Replace CheckItemExists with NativeItemExists

iSazonov committed:
Update NativeFileExists and NativeDirectoryExists

iSazonov committed:
Remove old NativeItemExists

iSazonov committed:
[Feature] Remove 'Native' from method names

NativeItemExists -> ItemExists
NativeFileExists -> FileExists
NativeDirectoryExists -> DirectoryExists

iSazonov commented:
Reopen to restart CIs.

iSazonov commented:
@TravisEz13 Could you please help - I can not find why CIs failed?

iSazonov commented:
Feature tests failed with Pester not found. :confused:

SteveL-MSFT commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,49 @@ internal static bool IsAdministrator()
+            catch
+            {
+            }
+            return false;
```
should be

>return result;

iSazonov replied:
Fixed.

SteveL-MSFT commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,49 @@ internal static bool IsAdministrator()
+            bool result = false;
+            try
+            {
+                result =  ItemExists(path, out bool unused);
```
extra whitespace after equals sign

SteveL-MSFT replied:
perhaps use named parameter to make it more clear what `unused` is?

iSazonov replied:
`unused` here is self explaination - we don't use the parameter at all.

iSazonov replied:
Extra space fixed.

SteveL-MSFT commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,49 @@ internal static bool IsAdministrator()
+            return false;
         }
 
-        // This is done through P/Invoke since File.Exists and Directory.Exists pay 13% performance degradation
```
Are we sure we no longer have a perf penalty with this change?

iSazonov replied:
We have still a perf penalty (on windows because normalization(!) - one extra kernel call in .Net code) but it is tracked in CoreFX https://github.com/dotnet/corefx/issues/19717
We have already mentioned that we do up to 7 the normalizations of the same path per cmdlet call. So perf penalty is more in PowerShell code. I'm trying to find how we can reduce the path normalization perf penalty. 

Also please see https://github.com/dotnet/corefx/issues/27049#issuecomment-392826352

iSazonov committed:
Fix return in ItemExists

iSazonov commented:
@TravisEz13 Thanks! But CIs failed again. I rebased - perhaps it resolve..

iSazonov committed:
Fix module loading

No need in throw here if file not found. (Previously we could throw in rare cases.)

iSazonov committed:
Recover ItemExists() behavior for File not found exception

daxian-dbw commented on src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs:
```diff
@@ -4111,7 +4111,7 @@ private ExternalScriptInfo FindLocalizedModuleManifest(string path)
                 String filePath = stringBuilder.ToString();
 
-                if (Utils.NativeFileExists(filePath))
+                if (File.Exists(filePath))
```
Why not use `Utils.FileExists` ?

iSazonov replied:
Utils.FileExists behavior is the same (no exception is thrown). So it is better to use the standard method.

daxian-dbw replied:
It's better to make all the `Utils.NativeFileExists -> File/Directory.Exists` changes in a separate PR. So for now, just do the simple replacement of `Utils.NativeFileExists -> Utils.FileExists`.

daxian-dbw commented on src/System.Management.Automation/engine/Modules/ModuleUtils.cs:
```diff
@@ -233,7 +233,7 @@ internal static IEnumerable<string> GetDefaultAvailableModuleFiles(string topDir
                 {
                     string moduleFile = Path.Combine(directoryToCheck, proposedModuleName) + ext;
-                    if (!Utils.NativeFileExists(moduleFile))
+                    if (!File.Exists(moduleFile))
```
Same here.
Actually, I see `Utils.NativeFileExists` and `File.Exists` mixed in this file (possibly in that files too). Do we really care about the exception on failure in those cases?

iSazonov replied:
Utils.FileExists behavior is the same (no exception is thrown). So it is better to use the standard method.

Yes, we should review all these cases but there are dozens of them to do in this PR or in a new?

daxian-dbw replied:
Yeah. let's do it in a separate PR. For this PR, just do the simple rename replacement: `Utils.NativeFileExists -> Utils.FileExists`.

daxian-dbw commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,61 @@ internal static bool IsAdministrator()
         }
 
-        // This is done through P/Invoke since File.Exists and Directory.Exists pay 13% performance degradation
-        // through the CAS checks, and are terribly slow for network paths.
```
I'm fine with changing to `File.GetAttribute()` because `CAS` is not supported in .NET Core, so the 13% penalty is gone. So I think the biggest concern of using the .NET API is not there anymore.

However, can you check the `"network paths"` scenario?

iSazonov replied:
The code come from Windows PowerShell and I don't know why it existed. I could not find any way where we would have throw on "network path not found" and not throw on "local path not found".
If we continue migration from ItemExists to File.Exists/Directory.Exists perhaps we found the amazing case.
Even if it is necessary we should eliminate the exception and replace it with regular check.
Now I suppose it's contrived

daxian-dbw replied:
No, I was saying about the "terribly slow for network paths" part. Is "File.Exists and Directory.Exists" still terribly slow on network paths? I guess not as they all checking for attribute now, just like our implementation here. But better to check out.

iSazonov replied:
I see a delay only if I use a non-existent server name. I suppose this usual behavior with DNS name resolving.

daxian-dbw commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,61 @@ internal static bool IsAdministrator()
+            bool result = false;
+            try
+            {
+                result = ItemExists(path, out bool unused);
```
> ItemExists(path, out bool unused)

I suggest changing it to `ItemExists(path, out _)`, using the "discard" variable.

daxian-dbw replied:
And, maybe

```
internal static bool ItemExists(string path)
{
    try { return ItemExists(path, out _) } catch { }
    return false;
}
```

daxian-dbw commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,61 @@ internal static bool IsAdministrator()
-        // This is done through P/Invoke since we pay 13% performance degradation
-        // through the CAS checks required by File.Exists and Directory.Exists
-        internal static bool NativeFileExists(string path)
+        internal static bool FileExists(string path)
```
I don't think all uses of this method expects an exception to be thrown. Some uses might just believe that it's faster than `File.Exists` but expect it to behave like `File.Exists()`.

I think you can inspect the uses of `Utils.FileExists` and `Utils.DirectoryExists` and change to `File.Exists` and `Directory.Exists` as appropriate.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -2921,15 +2891,15 @@ private void RemoveDirectoryInfoItem(DirectoryInfo directory, bool recurse, bool
+                try
                 {
-                    return;
+                    if (!Utils.ItemExists(directory.FullName, out bool unused))
```
Use the "discard" variable.

iSazonov replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -2921,15 +2891,15 @@ private void RemoveDirectoryInfoItem(DirectoryInfo directory, bool recurse, bool
-                if (accessException != null)
+                catch (Exception accessException)
                 {
                     ErrorRecord errorRecord = new ErrorRecord(accessException, "RemoveFileSystemItemUnAuthorizedAccess", ErrorCategory.PermissionDenied, directory);
```
Looks to me it's expecting an `UnauthorizedAccessException`, shall we narrow the catch to only catch this type of exception?

iSazonov replied:
Here I only preserve the logic.
Really we should clean up the code too because .Net Core support reparse points and recursion https://github.com/dotnet/corefx/blob/f25eb288a449010574a6e95fe298f3ad880ada1e/src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs#L359

Seems we should clean up all code in FileSystemProvider - perhaps we don't need FileSystemProviderV2 :-)

iSazonov replied:
@daxian-dbw I opened #7017 to clean up the code.

daxian-dbw replied:
I'm fine leaving this catch as is.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -3249,19 +3219,7 @@ private bool ItemExists(string path, out ErrorRecord error)
-                {
-                    throw accessException;
-                }
+                result = Utils.ItemExists(path, out bool unused);
```
"discard" variable.

iSazonov replied:
Fixed.

daxian-dbw commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -137,10 +137,10 @@ private static void EncodingRegisterProvider()
             string sysRoot = System.Environment.GetEnvironmentVariable("SystemRoot");
             string urlmonPath = Path.Combine(sysRoot, @"System32\urlmon.dll");
-            if (Utils.NativeFileExists(urlmonPath))
+            if (Utils.FileExists(urlmonPath))
```
At least for these two uses, they do not expect exception to be thrown. So change them to `File.Exists` should be fine.

iSazonov replied:
Fixed.

daxian-dbw replied:
Maybe revert these two changes too, since we agreed to use a separate PR for the migration changes.

iSazonov replied:
Reverted.

daxian-dbw commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,61 @@ internal static bool IsAdministrator()
-        // This is done through P/Invoke since we pay 13% performance degradation
-        // through the CAS checks required by File.Exists and Directory.Exists
-        internal static bool NativeDirectoryExists(string path)
+        internal static bool DirectoryExists(string path)
```
I think we should change the name of this one and `FileExists` to something that indicate it might throw exception, so it's easy to tell whether you should use them instead of `File.Exists` and `Directory.Exists` when writing code.

I suggest rename them to `CheckFileExistsAndThrowOnError` and `CheckDirectoryExistsAndThrowOnError`.

iSazonov replied:
I'll rename if we will not migrate to File.Exists/Directory.Exists in the PR.

daxian-dbw replied:
No, let's use a separate PR to do the migration changes. Go ahead rename this method.

daxian-dbw commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,61 @@ internal static bool IsAdministrator()
-        // This is done through P/Invoke since File.Exists and Directory.Exists pay 13% performance degradation
-        // through the CAS checks, and are terribly slow for network paths.
-        internal static bool NativeItemExists(string path, out bool isDirectory, out Exception exception)
+        internal static bool ItemExists(string path, out bool isDirectory)
```
Similarly, maybe changing this to `CheckItemExistsAndThrowOnError`.

iSazonov replied:
I'll rename if we will not migrate to File.Exists/Directory.Exists in the PR.

iSazonov commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,61 @@ internal static bool IsAdministrator()
-                    exception = new UnauthorizedAccessException(win32Exception.Message, win32Exception);
-                }
-                else if (errorCode == 32)
-                {
```
Please see the PR description (Fix is ready in [CoreFX](https://github.com/dotnet/corefx/issues/29782), we get it in 2.1.1.)

iSazonov replied:
Please see the PR description - we'll get the fix in .Net Core 2.1.1.

daxian-dbw replied:
OK, so this will only affect `c:\pagefile.sys`?

iSazonov replied:
Three files - pagefile.sys, hiberfil.sys, swapfile.sys. In the CoreFX issue a comment says that they can not emulate the behavior and so use the real file c:\pagefile.sys in test. I think users are unlikely to find something like this

iSazonov commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,61 @@ internal static bool IsAdministrator()
-#else
-
+#if !UNIX
             if (IsReservedDeviceName(path))
```
Seems method in line 898 (internal static bool ItemExists(string path)) is used on Unix.

daxian-dbw replied:
Take a look at the method `IsReservedDeviceName`, it's body already takes care of the `#if/def`.

iSazonov replied:
:-) I looked the other way.
Fixed.

iSazonov commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -895,113 +895,61 @@ internal static bool IsAdministrator()
+                // Use 'File.GetAttributes()' because we want to get access exceptions.
+                // TODO: we should review the tricky logic
+                // (historically we throw 'UnauthorizedAccessException' here) and migrate
+                // to standard methods 'File.Exist()' and 'Directory.Exists()' where possible.
```
Seems my comment is not clear.
I meant that we should migrate from the  ItemExists method to File.Exists() and Directory.Exits() which don't throw UnauthorizedAccessException.

daxian-dbw replied:
Yeah, I agree. I also pointed that out in my comments. Open an issue to track that migration instead of this comment. This comment is not useful in the context of this method, which is designed to throw on error.

iSazonov replied:
Done #7031

iSazonov replied:
TODO comment was removed.

iSazonov committed:
Use "discard" variable

iSazonov committed:
Use File.Exists in GetFileSecurityZone

daxian-dbw commented:
@iSazonov No, you don't need to do it in this PR. Open an issue to track the analysis and migration work (Utils.FileExists/DirectoryExists => File.Exists/Directory.Exists).

iSazonov commented:
Tracking Issue #7031

iSazonov committed:
Remove unneeded #if !UNIX

iSazonov committed:
Revert "Use File.Exists in GetFileSecurityZone"

This reverts commit db25895ef801dbb64b49c891322624fa9bf49f30.

iSazonov committed:
Remove unneeded comment

iSazonov commented:
@daxian-dbw File.Exists() and Directory.Exists() return false for device names so we can remove IsReservedDeviceName call from ItemExists() method at all. What do you think?

daxian-dbw commented:
@iSazonov `ItemExists` doesn't call `FileExists` or `Directory.Exists`. Do you mean `File.GetAttributes` returns -1 for reserved device names?

daxian-dbw commented:
The changes LGTM.
@iSazonov Are you going to rename `FileExists`, `DirectoryExists` and `ItemExists` to
`CheckFileExistsAndThrowOnError`, `CheckDirectoryExistsAndThrowOnError` and `CheckItemExistsAndThrowOnError` in this PR?

iSazonov commented:
> Are you going to rename FileExists, DirectoryExists and ItemExists to
CheckFileExistsAndThrowOnError, CheckDirectoryExistsAndThrowOnError and CheckItemExistsAndThrowOnError in this PR?

No, I'll remove the methods in follow PR.

>ItemExists doesn't call FileExists or Directory.Exists. Do you mean File.GetAttributes returns -1 for reserved device names?

Never mind - I hope to remove ItemExists() altogether in follow PR.

iSazonov commented:
@daxian-dbw Should I fix something else?

daxian-dbw reviewed (approved):
@iSazonov If you plan to remove those methods in your next PR, then it's fine to leave their names unchanged.

daxian-dbw commented:
@iSazonov It's preferred that you remove the native code from psl-native along with your managed code changes in your following PRs. For the PRs you already finished, I listed the tasks to track removing the corresponding native code in #6874.

iSazonov commented:
@daxian-dbw My thoughts was that if we have to re-publish psl-native after every change 
it is better to finish first the cleanup in the repo and only then cleanup in psl-native repo.
